### PR TITLE
Add support for DirectCounter extern to PSA/eBPF backend

### PR DESCRIPTION
**Pull request by tatry**
This PR adds the support for DirectCounter extern to PSA/eBPF backend.

Co-authored-by: Tomasz Osiński <tomasz@opennetworking.org>
Co-authored-by: Mateusz Kossakowski <mateusz.kossakowski@orange.com>

### CODE REVIEW

File: backends/ebpf/ebpfTable.h [hunk 1]
```diff
@@ -104,6 +104,7 @@ class EBPFTable : public EBPFTableBase {
     virtual void emitValueType(CodeBuilder* builder);
     virtual void emitValueActionIDNames(CodeBuilder* builder);
     virtual void emitValueStructStructure(CodeBuilder* builder);
+    virtual void emitDirectTypes(CodeBuilder* builder) { (void) builder; }
```
Comment (mbudiu-vmw) [inline backends/ebpf/ebpfTable.h:107]: can you document this function? I can't guess what it's doing from its name.

File: backends/ebpf/psa/ebpfPsaTable.cpp [hunk 2]
```diff
@@ -23,12 +23,36 @@ limitations under the License.
 
 namespace EBPF {
 
+template<class F>
+class EBPFTablePsaPropertyVisitor : public Inspector {
+ protected:
+    F functor;  // called for every property entry
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/ebpfPsaTable.cpp:29]: seems to be called for their values which are PathExpression only.

File: backends/ebpf/psa/ebpfPsaTable.cpp [hunk 3]
```diff
@@ -54,16 +78,14 @@ void ActionTranslationVisitorPSA::processMethod(const P4::ExternMethod* method)
...
+    if (declType->name.name == "DirectCounter") {
+        auto ctr = table->getDirectCounter(instanceName);
+        if (ctr != nullptr)
+            ctr->emitDirectMethodInvocation(builder, method, valueName);
+        else
             ::error(ErrorType::ERR_NOT_FOUND,
-                    "%1%: Counter named %2% not found",
-                    method->expr, instanceName);
-        }
+                    "%1%: Table %2% do not own DirectCounter named %3%",
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/ebpfPsaTable.cpp:87]: "does"

File: backends/ebpf/psa/ebpfPsaTable.cpp [hunk 4]
```diff
@@ -175,6 +213,17 @@ void EBPFTablePSA::emitTypes(CodeBuilder* builder) {
     // TODO: placeholder for handling PSA-specific types
 }
 
+/**
+ * Remember that order of emitting counters and meters affects future access to BPF maps.
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/ebpfPsaTable.cpp:217]: I find this comment mysterious. "future"? In which way does it affect the access?

File: backends/ebpf/psa/externs/ebpfPsaCounter.cpp [hunk 5]
```diff
@@ -153,6 +187,35 @@ void EBPFCounterPSA::emitMethodInvocation(CodeBuilder* builder, const P4::Extern
...
 }
 
+void EBPFCounterPSA::emitDirectMethodInvocation(CodeBuilder* builder,
+                                                const P4::ExternMethod* method,
+                                                cstring valuePtr) {
+    if (method->method->name.name != "count") {
+        ::error(ErrorType::ERR_UNSUPPORTED, "Unexpected method %1%", method->expr);
+        return;
+    }
+    BUG_CHECK(isDirect, "Bad Counter invocation");
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/externs/ebpfPsaCounter.cpp:197]: should these be ::error, or are you satisfied with BUG?

**Review by mbudiu-vmw — APPROVED** [review]
I am approving, but I have a few small comments.

File: backends/ebpf/psa/ebpfPsaTable.cpp (hunk 2 above)
Comment (tatry) [reply backends/ebpf/psa/ebpfPsaTable.cpp:29]: What do you mean? The comment should be clarified or something else?

File: backends/ebpf/psa/externs/ebpfPsaCounter.cpp (hunk 5 above)
Comment (tatry) [reply backends/ebpf/psa/externs/ebpfPsaCounter.cpp:197]: This situation should never occur, where for DirectCounter emitMethodInvocation is called and for Counter emitDirectMethodInvocation is called. This has more to do with compiler implementation than P4 code, so `BUG` is a better option in my opinion.

File: backends/ebpf/psa/ebpfPsaTable.cpp (hunk 2 above)
Comment (mbudiu-vmw) [reply backends/ebpf/psa/ebpfPsaTable.cpp:29]: Actually the right way to do this is to have this be a base class and define a new subclass for each different analysis, with just a preorder(PathExpression*) method. Please rewrite it that way.

**Comment by tatry** [comment]
@mbudiu-vmw thank you for your review, I think that I've addressed all the comments.

File: backends/ebpf/psa/ebpfPsaTable.cpp [hunk 6]
```diff
@@ -23,12 +23,106 @@ limitations under the License.
...
 
+class EBPFTablePsaPropertyVisitor : public Inspector {
+ protected:
+    EBPFTablePSA* table;
+
+ public:
+    explicit EBPFTablePsaPropertyVisitor(EBPFTablePSA* table) : table(table) {}
+
+    // Use these two preorders to print error when property contains something other than name of
+    // extern instance. ListExpression is required because without it Expression will take precede
```
Comment (mbudiu-vmw) [inline backends/ebpf/psa/ebpfPsaTable.cpp:34]: precedence

**Review by mbudiu-vmw — APPROVED** [review]

**Pull request merged**